### ISSUE TITLE
Allow more generic log_events

### ISSUE THREAD

**Issue by jchuong** [Netflix/lemur#984]
````
class Log(db.Model):
    __tablename__ = 'logs'
    id = Column(Integer, primary_key=True)
    certificate_id = Column(Integer, ForeignKey('certificates.id'))
    log_type = Column(Enum('key_view', 'create_cert', 'update_cert', 'revoke_cert', name='log_type'), nullable=False)
    logged_at = Column(ArrowType(), PassiveDefault(func.now()), nullable=False)
    user_id = Column(Integer, ForeignKey('users.id'), nullable=False)
````

Essentially, if I want to write a plugin that writes to log_service, then I will have to modify the log_type in Lemur core code, instead of having Lemur find possible log_types in any active plugin.

**Comment by kevgliss** [comment]
What kind of events where you looking to log? The events log was intended to be an audit log for potentially sensitive operations.

**Comment by jchuong** [comment]
Plugin-related messages, I suppose.  I have a custom Destination plugin, but this applies to every destination.  Essentially, there is currently no way to know if an `upload()` to a destination has succeeded or failed.  Since uploading to a destination is essentially updating a production machine, it would be nice to let an operator know if/when a certificate upload failed so we can troubleshoot.  Another scenario would be an operator uploading the wrong certificate to the wrong destination (say, added a destination to the list), and we can trace that event back.

 Related note: `upload()` can raise an exception, but is only logged back into lemur.log, instead of alerting the operator that the creation/update of a certificate failed.
````
@event.listens_for(Certificate.destinations, 'append')
def update_destinations(target, value, initiator):
    """
    Attempt to upload certificate to the new destination

    :param target:
    :param value:
    :param initiator:
    :return:
    """
    destination_plugin = plugins.get(value.plugin_name)

    try:
        if target.private_key:
            destination_plugin.upload(target.name, target.body, target.private_key, target.chain, value.options)
    except Exception as e:
        current_app.logger.exception(e)
        metrics.send('destination_upload_failure', 'counter', 1,
                     metric_tags={'certificate': target.name, 'destination': value.label})
        # re-raise issue so user can see that plugin failed to upload, instead of failing silently
        raise e
````
I have re-raised the exception so that the certificate update fails.  This is a problem since there is also currently no way to re-push a certificate to their destination(s). If the update succeeds, and the upload to destination fails, then the current implementation will silently update the cert in the DB, but the operator will not know if anything happened.  Then, the only way to re-upload is to modify the destinations again, to force an upload. 

This is another event that probably should be logged.  Raising the exception keeps the operator from being able to commit the updated cert into the DB.

**Comment by kevgliss** [comment]
Yes, re-doing the destination lifecycle has been a problem for quite awhile. When thinking about a solution I was having a problem in determining what success would be?

For instance when creating a certificate with a destination, if the certificate creation succeeds but a destination fails, is an error returned to the user? Would that confuse them as they get a certificate back?

Ultimately I think that those type of operations would need to be moved to a "tasks" table. Destination uploads, certificate rotations, etc., would be tasks linked to the certificate in question. In this way we could de-couple the issuance aspects from the deployment aspects. 

Unfortunately, it's not a huge priority for us atm so I can't really comment on when I might get to work on such a feature. 

If you're interested in implementing such a feature I would be more than happy to review and provide guidance.

**Comment by jchuong** [comment]
Hmm, well I can possibly create a way to detect log types. 

For destination uploading (force uploading), I will probably end up writing that endpoint / adding to UI at some point in the near future.  That may also take out auto-upload on updates, which may flow into your task lifecycle.

As for the larger "task table", I'm probably not going to be able to do that at the moment.  I'm not entirely sure of your direction just yet and I'm still working on using Lemur as part of our workflow for now.

**Comment by jchuong** [comment]
Another point to this issue, logging is directly tied to a certificate.  It would also be nice if that didn't have to be the case (ex. Plugin export right now, does not have a reference to a certificate, but just body, chain, and private key.  The plugin wouldn't be able to write to log_service themselves if there was an issue, but only certificate/service.py could with a generic message.
